File: lib/src/rules/security/file_upload_validation_rule.dart
import '../../models/finding.dart';
import '../../models/project_context.dart';
import '../../rule.dart';

class FileUploadValidationRule extends Rule {
  const FileUploadValidationRule();

  @override
  String get code => 'file-upload-validation';

  @override
  List<Finding> evaluate(ProjectContext context) {
    final findings = <Finding>[];

    for (final upload in context.uploadCalls) {
      if (upload.hasValidationHelper ||
          (upload.hasTypeValidation && upload.hasSizeValidation)) {
        continue;
      }

      final missingChecks = <String>[];
      if (!upload.hasTypeValidation) {
        missingChecks.add('file type');
      }
      if (!upload.hasSizeValidation) {
        missingChecks.add('size');
      }

      findings.add(
        Finding(
          severity: FindingSeverity.medium,
          confidence: FindingConfidence.medium,
          category: FindingCategory.security,
          code: code,
          message:
              'Upload call is missing nearby ${missingChecks.join(' and ')} validation',
          fix:
              'Validate MIME/extension and file size before calling `.upload(...)` so unsafe files are rejected on the client before they hit storage.',
          risk:
              'Unvalidated client uploads can lead to malicious file execution, oversized payloads, or storage quota exhaustion.',
          filePath: upload.file.relativePath,
          line: upload.line,
        ),
      );
    }

    return findings;
  }
}

File: vscode-extension/src/rules/security/fileUploadValidationRule.ts
import { Rule } from '../rule';
import { Finding, FindingSeverity, FindingCategory, FindingConfidence, DetectionMethod } from '../../models/finding';
import { ProjectContext } from '../../scanner/projectContext';

export class FileUploadValidationRule implements Rule {
  readonly code = 'file-upload-validation';

  evaluate(context: ProjectContext): Finding[] {
    const findings: Finding[] = [];
    for (const upload of context.uploadCalls) {
      if (upload.hasValidationHelper || (upload.hasTypeValidation && upload.hasSizeValidation)) {
        continue;
      }
      const missing: string[] = [];
      if (!upload.hasTypeValidation) { missing.push('file type'); }
      if (!upload.hasSizeValidation) { missing.push('size'); }
      findings.push(new Finding({
        severity: FindingSeverity.medium,
        confidence: FindingConfidence.medium,
        detectionMethod: DetectionMethod.structural,
        category: FindingCategory.security,
        code: this.code,
        message: `Upload call is missing nearby ${missing.join(' and ')} validation`,
        fix: 'Validate MIME/extension and file size before calling `.upload(...)` so unsafe files are rejected on the client before they hit storage.',
        risk: 'Unvalidated client uploads can lead to malicious file execution, oversized payloads, or storage quota exhaustion.',
        filePath: upload.file.relativePath,
        line: upload.line,
      }));
    }
    return findings;
  }
}
